source/h1_connection.c

```diff
@@ -381,49 +306,45 @@ int aws_h1_stream_activate(struct aws_http_stream *stream) {
             return AWS_OP_SUCCESS;
         }
 
-        err = connection->synced_data.new_stream_error_code;
-        if (err) {
+        if (connection->synced_data.new_stream_error_code) {
             aws_h1_connection_unlock_synced_data(connection);
-            goto error;
+            AWS_LOGF_ERROR(
+                AWS_LS_HTTP_CONNECTION,
+                "id=%p: Failed to activate the stream id=%p, new streams are not allowed now. error %d (%s)",
+                (void *)&connection->base,
+                (void *)stream,
+                connection->synced_data.new_stream_error_code,
+                aws_error_name(connection->synced_data.new_stream_error_code));
+            return aws_raise_error(connection->synced_data.new_stream_error_code);
         }
 
         stream->id = aws_http_connection_get_next_stream_id(base_connection);
+        if (!stream->id) {
+            aws_h1_connection_unlock_synced_data(connection);
+            /* aws_http_connection_get_next_stream_id() raises its own error. */
+            return AWS_OP_ERR;
+        }
```

Comment:
the error handling code-flow in `activate_stream()` bothered me, it was all over the place

now it always unlocks and returns early, which I usually don't approve of, but it was the cleanest thing I could come up with